cirq-google/cirq_google/devices/grid_device.py

```diff
@@ -0,0 +1,274 @@
+# Copyright 2022 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Device object representing Google devices with a grid qubit layout."""
+
+from typing import Any, Set, Tuple, cast
+import cirq
+from cirq_google.api import v2
+
+
+def _validate_device_specification(proto: v2.device_pb2.DeviceSpecification) -> None:
+    """Validates the DeviceSpecification proto.
+
+    Args:
+        proto: The DeviceSpecification proto to validate.
+
+    Raises:
+        ValueError: If the DeviceSpecification is invalid.
+
+    """
+
+    for target_set in proto.valid_targets:
+
+        # Check for unknown qubits in targets.
+        for target in target_set.targets:
+            for target_id in target.ids:
+                if target_id not in proto.valid_qubits:
+                    raise ValueError(
+                        f"Invalid DeviceSpecification: valid_targets contain qubit '{target_id}'"
+                        " which is not in valid_qubits."
+                    )
+
+        # Symmetric and asymmetric targets should not have repeated qubits.
+        if (
+            target_set.target_ordering == v2.device_pb2.TargetSet.SYMMETRIC
+            or target_set.target_ordering == v2.device_pb2.TargetSet.ASYMMETRIC
+        ):
+            for target in target_set.targets:
+                if len(target.ids) > len(set(target.ids)):
+                    raise ValueError(
+                        f"Invalid DeviceSpecification: the target set '{target_set.name}' is either"
+                        " SYMMETRIC or ASYMMETRIC but has a target which contains repeated qubits:"
+                        f" {target.ids}."
+                    )
+
+        # A SUBSET_PERMUTATION target should contain exactly one qubit.
```

Comment:
I don't think this is true.  Why does a subset permutation contain exactly one qubit?

If I recall, a subset permutation means that you can have any subset of the qubits.  This is used for gates like measurement where you can measure any subset of the qubits.

Reply:
SUBSET_PERMUTATION describes a *target set* rather than a target, where one can have any subset of the targets, with each target being exactly 1 qubit.

https://github.com/quantumlib/Cirq/blob/4ff626477bd93a81285c5673cc56044822eaa8af/cirq-google/cirq_google/api/v2/device.proto#L154-L159

Reply:
Ah, I see.  Maybe link to this proto in the comment or otherwise clarify, since it is quite confusing to remember how this is done otherwise.